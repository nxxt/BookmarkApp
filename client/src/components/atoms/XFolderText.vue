<template>
  <div class="x-folder-text d-flex align-center">
    <v-icon class="mr-1" x-small>mdi-folder</v-icon>
    <span
      v-for="(item, i) in folders"
      :key="item._id"
      :class="{ 'mr-1': i !== folders.length - 1 }"
      @click="$emit('click', item._id)"
    >
      {{ `${item.name}${i === items.length - 1 ? '' : ','}` }}
    </span>
  </div>
</template>

<script lang="ts">
import { PropType } from 'vue'
import { Component, Prop, Vue } from 'vue-property-decorator'
import IFolder from '@@/types/IFolder'

@Component
export default class XFolderText extends Vue {
  @Prop({ type: Array as PropType<String[]>, default: () => [] })
  items!: string[]

  get folders(): IFolder[] {
    return this.items
      .map((id) => this.$getFolderById(id))
      .sort((a, b) => (a.name > b.name ? 1 : -1))
  }
}
</script>
