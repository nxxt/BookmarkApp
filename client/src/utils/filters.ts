import IBookmark from '~~/types/IBookmark'
import ITag from '~~/types/ITag'
import IFolder from '~~/types/IFolder'

const filters: { [key: string]: Function } = {
  bookmarkTitle: (bookmark: IBookmark) => {
    return bookmark && bookmark.title ? bookmark.title : ''
  },
  tagName: (tag: ITag) => {
    return tag && tag.name ? `#${tag.name}` : ''
  },
  folderName: (folder: IFolder) => {
    return folder && folder.name ? folder.name : ''
  },
  textOverflow: (str: string) => {
    return str.length > 140 ? `${str.substring(0, 140)}â€¦` : str
  },
  dateFormat: (str: string) => {
    return str.substring(0, 10)
  }
}

export default filters
