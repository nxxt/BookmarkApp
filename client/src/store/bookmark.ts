import { Module, VuexModule, Mutation, Action } from 'vuex-module-decorators'
import Bookmark from '@@/types/IBookmark'
// import DtoBookmark from '@@/types/DtoBookmark'
import ISortType from '@@/types/ISortType'
import ITag from '@@/types/ITag'
import { $axios } from '@/utils/api'
// import IFolder from '~~/types/IFolder'

// stateFactory: true → Vuex をモジュールモードで扱うために指定
@Module({
  stateFactory: true,
  namespaced: true,
  name: 'bookmark'
})
export default class Bookmarks extends VuexModule {
  private all: Bookmark[] = []
  private unfoldered: Bookmark[] = []
  private display: Bookmark[] = []
  private folderId: string = ''

  get allBookmarks() {
    return this.all
  }

  get unfolderedBookmarks() {
    return this.unfoldered
  }

  get displayBookmarks() {
    return this.display
  }

  get _fId() {
    return this.folderId
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private FETCH_ALL(items: Bookmark[]) {
    this.all = items
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private FETCH_UNFOLDERED(items: Bookmark[]) {
    this.unfoldered = items
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private FETCH_DISPLAY(items: Bookmark[], folderId: string) {
    this.display = items
    this.folderId = folderId
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private ADD_DISPLAY(items: Bookmark[]) {
    this.display = [...this.display, ...items]
  }

  /**
   * Bookmark を追加する
   * @param bookmark Bookmark オブジェクト
   */
  @Mutation
  private CREATE(data: { bookmark: Bookmark; folderId: string }) {
    // リストに Bookmark を追加
    this.all.push(data.bookmark)
    this.unfoldered.push(data.bookmark)
    // 選択したsidemenuが「all bookmarks」「unfoldered」の場合
    if (['h-0', 'h-1'].includes(data.folderId)) {
      this.display.push(data.bookmark)
    }
  }

  /**
   * Bookmark を更新する
   * @param bookmark Bookmark オブジェクト
   */
  @Mutation
  private UPDATE(bookmark: Bookmark) {
    const target = this.all.find((d) => d._id === bookmark._id)
    if (target) {
      const index = this.all.indexOf(target)
      Object.assign(this.all[index], bookmark)
    }
    const target2 = this.unfoldered.find((d) => d._id === bookmark._id)
    if (target2) {
      const index = this.unfoldered.indexOf(target2)
      Object.assign(this.unfoldered[index], bookmark)
    }
    const target3 = this.display.find((d) => d._id === bookmark._id)
    if (target3) {
      const index = this.display.indexOf(target3)
      Object.assign(this.display[index], bookmark)
    }
  }

  /**
   * Bookmark を削除する
   * @param bookmark 削除する Bookmark インスタンス
   */
  @Mutation
  private DELETE(_id: string) {
    const target = this.all.find((d: Bookmark) => d._id === _id)
    if (target) {
      const index = this.all.indexOf(target)
      this.all.splice(index, 1)
    }
    const target2 = this.unfoldered.find((d: Bookmark) => d._id === _id)
    if (target2) {
      const index = this.unfoldered.indexOf(target2)
      this.unfoldered.splice(index, 1)
    }
    const target3 = this.display.find((d: Bookmark) => d._id === _id)
    if (target3) {
      const index = this.display.indexOf(target3)
      this.display.splice(index, 1)
    }
  }

  @Action({})
  public async fetchAllBookmarks(params: { userId: string }): Promise<void> {
    const data = await $axios.$get(`/api/v1/bookmarks/all/${params.userId}`)
    this.FETCH_ALL(data)
  }

  @Action({})
  public async fetchUnfolderedBookmarks(params: {
    userId: string
  }): Promise<void> {
    const data = await $axios.$get(
      `/api/v1/bookmarks/unfoldered/${params.userId}`
    )
    this.FETCH_UNFOLDERED(data)
  }

  @Action({})
  public async fetchDisplayBookmarks(params: {
    userId: string
    pageNo: number
    perPage: number
    folderId: string
    filter: (string | ITag)[]
    sort: ISortType
  }): Promise<any> {
    const { userId, pageNo, perPage, folderId, filter, sort } = params
    const res = await $axios.post(
      `/api/v1/bookmarks/list/${userId}?page_no=${pageNo}&per_page=${perPage}&folder_id=${folderId}`,
      { sort, filter }
    )
    this.FETCH_DISPLAY(res.data, params.folderId)
    return [res.data, Number(res.headers['x-list-total-length'])]
  }

  @Action({})
  public async addDisplayBookmarks(params: {
    userId: string
    pageNo: number
    perPage: number
    folderId: string
    filter: (string | ITag)[]
    sort: ISortType
  }): Promise<void> {
    const { userId, pageNo, perPage, folderId, filter, sort } = params

    const data = await $axios.$post(
      `/api/v1/bookmarks/list/${userId}?page_no=${pageNo}&per_page=${perPage}&folder_id=${folderId}`,
      { sort, filter }
    )
    this.ADD_DISPLAY(data)
  }

  @Action({})
  async createBookmark(data: {
    url: string
    userId: string
    folderId: string
  }) {
    const item = await $axios.$post(`/api/v1/bookmarks`, {
      url: data.url,
      userId: data.userId
    })
    this.CREATE({ bookmark: item, folderId: data.folderId })
    return item
  }

  @Action({})
  async updateBookmark(bookmark: Bookmark) {
    const item = await $axios.$patch(
      `/api/v1/bookmarks/${bookmark._id}`,
      bookmark
    )
    this.UPDATE(item)
    return item
  }

  @Action({})
  async deleteBookmark(_id: string) {
    const data = await $axios.$delete(`/api/v1/bookmarks/${_id}`)
    this.DELETE(_id)
    return data
  }
}
