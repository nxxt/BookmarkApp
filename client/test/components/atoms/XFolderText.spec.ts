import Vuetify from 'vuetify'
import { mount, createLocalVue } from '@vue/test-utils'
import bookmark from '@@/test/mock_data/bookmark.json'
import filters from '@/utils/filters'
import mockGetters from '@/utils/mock_getters'
import XFolderText from '@/components/atoms/XFolderText.vue'

// filter, plugin 登録
const localVue = createLocalVue()
Object.keys(filters).forEach((key: string) =>
  localVue.filter(key, filters[key])
)
Object.keys(mockGetters).forEach((key: string) => {
  localVue.prototype[key] = mockGetters[key]
})

describe('XFolderText.vue', () => {
  // vuetify の準備
  let vuetify: typeof Vuetify
  beforeEach(() => {
    vuetify = new Vuetify()
  })

  // propsで渡す値の準備
  const propsData = {
    items: bookmark.folder // folderのobjectId配列
  }

  const factory = (options?: any) => {
    return mount(XFolderText, {
      localVue,
      vuetify,
      ...options
    })
  }

  /* const translate = (_id: string) => {
    return filters.folderName(mockGetters.$getFolderById(_id))
  } */

  // TEST
  it('props', () => {
    const wrapper = factory({ propsData })
    expect(wrapper.props()).toStrictEqual(propsData)
  })

  // TODO: spanのtextが取得できない
  /* describe('template parts', () => {
    it('text', () => {
      const wrapper = factory({ propsData })
      const elms = wrapper.findAll('span').wrappers
      elms.forEach((span, i) => {
        if (i === elms.length - 1) {
          console.log(span.text())
          expect(span.text()).toBe(`${translate(propsData.items[i])},`)
        } else {
          console.log(span.text())
          expect(span.text()).toBe(translate(propsData.items[i]))
        }
      })
      const elm = wrapper.find('.x-folder-text span')
      expect(elm.text()).toBe(translate(propsData.items[0]))
    })
  }) */

  describe('the entire template', () => {
    it('snapshot', () => {
      const wrapper = factory({ propsData })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
