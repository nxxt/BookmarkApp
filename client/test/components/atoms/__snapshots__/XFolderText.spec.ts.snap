// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XFolderText.vue the entire template snapshot 1`] = `
"<div class=\\"x-folder-text d-flex align-center\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mr-1 mdi mdi-folder theme--light\\" style=\\"font-size: 12px;\\"></i> <span class=\\"mr-1\\">
    FOLDER001,
  </span><span class=\\"\\">
    FOLDER002
  </span></div>"
`;
